libraries/standard/mqtt/src/mqtt.c

```diff
@@ -191,10 +506,49 @@ MQTTStatus_t MQTT_Disconnect( MQTTContext_t * const pContext )
     return status;
 }
 
-MQTTStatus_t MQTT_Process( MQTTContext_t * const pContext,
-                           uint32_t timeoutMs )
+MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * const pContext,
+                               uint32_t timeoutMs )
 {
-    return MQTTSuccess;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTGetCurrentTimeFunc_t getCurrentTime = pContext->callbacks.getTime;
+    uint32_t entryTime = getCurrentTime();
+    MQTTPacketInfo_t incomingPacket;
+
+    while( ( getCurrentTime() - entryTime ) < timeoutMs )
+    {
+        status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
+                                                      &incomingPacket );
+
+        if( ( status != MQTTSuccess ) && ( status != MQTTNoDataAvailable ) )
+        {
+            IotLogErrorWithArgs( "Receiving incoming packet length failed. Status=%d",
+                                 status );
+        }
+
+        /* Receive packet. */
+        if( status == MQTTSuccess )
+        {
+            status = _receivePacket( pContext, incomingPacket );
+        }
+
+        /* Handle received packet. */
+        if( status == MQTTSuccess )
+        {
+            incomingPacket.pRemainingData = pContext->networkBuffer.pBuffer;
+            status = _handleMessage( pContext, &incomingPacket );
+        }
+
+        if( status != MQTTSuccess && status != MQTTNoDataAvailable )
+        {
+            /* Error, break. */
+            IotLogErrorWithArgs( "Exiting receive loop. Error status=%d",
+                                 status );
+            /* TODO signal connection to close? */
```

Comment:
Here, MQTT_GetIncomingPacketTypeAndLength() would have returned one of the network error type.  For now, application can deal with various errors and decide to close and reconnect. When the library starts bubbling up error code, library should return only one error code and application can then look at bubbled up error code from the transport layer.

Reply:
Removed comment